### DISCUSSION TITLE
How to tackle complex, nested data structures with lists?

### DISCUSSION THREAD

**Discussion by SzczurekYT** [rrousselGit/riverpod#2837] [Q&A]
Hello
I have a pretty complex data structure and I fail to implement it with Riverpod.
What I have is node graph, basically a list of `Node`s.
```dart
class Node {
  final String name;
  final String id;
  final List<Port> ports = [];
  Offset position = Offset.zero;

  Node({required this.name, required this.id});
}
```
Each `Node` keeps a list of `Port`s that can be connected
```dart
class Port {
  final String id;
  final PortType type;
  final PortDirection direction;
  final Node parentNode;
  final List<Link> links = [];
  final GlobalKey widgetKey = GlobalKey();

  Port(this.id, this.type, this.parentNode);
}
```
And each Port keeps a list of `Link`s - connections between two ports.
```dart
class Link {
  final Port sourcePort;
  final Port targetPort;

  Link(this.sourcePort, this.targetPort);
}
```
How I should manage this with Riverpod the immutable way (as it's recommended)?
I tried to make something myself with `StateNotfier`s, but wasn't able to create anything usable, hence I'm asking.
For list I'm gonna use the [fast_immutable_collections](https://pub.dev/packages/fast_immutable_collections) package, so they are immutable and much easier to manipulate.

**Comment by rrousselGit** [comment]
This is bit broad/unclear. Do you have a specific question? Otherwise there's not much to say.

**Comment by SzczurekYT** [comment]
So while explaining my issue with more details I realized one thing I had wrong in my mind. I started changing my code and it fixed some issue, but not all of them. I will post them once I will finish the rewriting.

**Comment by SzczurekYT** [comment]
Sorry for being inactive but I was absent for some time.
The issue I had in my mind was that I thought in a mutable way. I thought that for example when you have a list of objects you will need to watch both the list (for changes in it) and then all the objects to get updates of object's proprieties. That would be the case with mutable objects, but it isn't with immutable ones - when you change something in object you need to replace the object in the list and to do that you need to replace the list, so you don't need to watch objects.
With this in mind I managed to rewrite my app in a way that works with Riverpod.
There is bit of mess as in some places I keep reference to object by reference and in some by it's id, but it kinda works.
I have a `NodeGraphNotifier` class which keeps a map of nodes by their id and has methods to update them.
But I have two issues:
1. Now I watch the entire graph at once, passing each node to a NodeWidget, that passes each Port to it's PortWidget.
I feel like this may cause many unnecesary rebuilds. For example PortWidget doesn't care about the list of Link's each Port keeps, but will get a rebuild since adding/removing a link updates the entire structure.
2. This problem might be related to the first one.
Before the node pos was an ephermal state, keept inside a widget, and update with hooks. But now it's part of a Node class (the position field), and it's managed by riverpod. That's how I want it, but now moving a node is stuttery and lags behind the mouse (the node should stick to the mouse cursor), as you can see:

https://github.com/rrousselGit/riverpod/assets/67533827/400ba9f2-65f7-464c-acf4-ae1378586024

The entire code is here in this [repo](https://github.com/SzczurekYT/WaveMixer/tree/ee9609b3f1bb53fb9a9787bea20527dd94bc6cfc), you can browse it. Use the commit from this link, as the next one (latest) is kinda unfinished.
So my question is: Is there a way to change my code to make node moving smooth and be sure there aren't any big amounts of unnecesary rebuilds?
